### DISCUSSION TITLE
Using latex `\newcommand` with revealjs output

### DISCUSSION THREAD

**Discussion by martintburgess** [quarto-dev/quarto-cli#2068] [Q&A] [answered]
Thanks for all your work on quarto - it's awesome. 

I am following the example [here](https://github.com/quarto-dev/quarto-web/blob/main/docs/presentations/revealjs/demo/index.qmd#L132) for including equations in a quarto document using the revealjs format. 

When I knit the following `.qmd` file the first slide has no output but the second slide has the desired output. Is there a recommended way to add latex code/equations to revealjs presents? Happy to be pointed to other examples - haven't found any.

```
---
title: "mre"
format: 
  revealjs:
    self-contained-math: true
---

## With newcommand, no output

```{=tex}
\newcommand\muy{\mathrm{mu\_y}}
\begin{equation}
\muy
\end{equation}
```

## Without newcommand, output generated

```{=tex}
\begin{equation}
\mathrm{mu\_y}
\end{equation}
```

```

**Comment by mcanouil** [comment]
Reveal.js is HTML, thus it raw tex is ignored.
Equations are a bit different because they are the only LaTeX commands processed with JavaScript: webtex, gladtex, mathml, mathjax or katex[^1].
To note, some of those tools have an extension mechanism that you can use, see <https://github.com/quarto-dev/quarto-cli/discussions/1620#discussioncomment-3296389> which uses a MathJax extension.

If you really want to use LaTeX, choose beamer, not Reveal.js: <https://quarto.org/docs/presentations/beamer.html>

[^1]: <https://quarto.org/docs/reference/formats/html.html#format-options>

**Comment by cderv** [comment]
@martintburgess 

HTML math in **revealjs** will use MathJax. In MathJax, if support exist for a TeX command, then you can use it. It happens that `\newcommand` is one of them (http://docs.mathjax.org/en/latest/input/tex/extensions/newcommand.html)

However, you need to use Math environment, with `$$`

````markdown
---
title: "mre"
format: 
  revealjs:
    self-contained-math: true
---

$$
\newcommand\muy{\mathrm{mu\_y}}
\begin{equation}
\muy
\end{equation}
$$
````


By using only Raw latex with `` ```{=tex} ``  you are explicitly telling Pandoc that the content is raw LaTeX and should be ignored in HTML output then. HTML writer does an exception for raw LaTeX that start with a math environment, but in your case that is not the case as `\newcommand` is before. If you put it inside, MathJax should now how to handle it. 

Example 

````markdown
---
title: "mre"
format: 
  revealjs:
    self-contained-math: true
---

```{=tex}
\begin{equation}
\newcommand\muy{\mathrm{mu\_y}}
\muy
\end{equation}
```
````

But this would work too 
````markdown
---
title: "mre"
format: 
  revealjs:
    self-contained-math: true
---

\begin{equation}
\newcommand\muy{\mathrm{mu\_y}}
\muy
\end{equation}

````

In Markdown using the `$$` syntax is easier, also because some IDE (like RStudio visual editor) will help for preview I think. 

Overall, Mathjax processing adds a layer that needs to be understood when some specific customization are desired. 

Hope it helps understand

**Comment by cscheid** [comment, marked as answer]
Note that (as I learned yesterday from another quarto user), \newcommand _does_ work. You just need to adjust your syntax:

````
---
format: revealjs
---

\newcommand\muy{\mu_y}

We use $\muy$ (@eq-muy) to denote the mean of the $y$ parameter:

$$\muy$$ {#eq-muy}
````

Specifically, I think the issue you're seeing (on top of raw latex not being emitted in html formats) is that in quarto, we typically don't use `\begin{equation}`, we use display math (`$$`) and regular crossreferences (`{#eq-muy}`).

If you're careful to use `\newcommand` outside math mode, then the markdown should be portable across formats. This way, you get the following slides by default:

<img width="891" alt="image" src="https://user-images.githubusercontent.com/285675/186194906-0cea35cc-a647-4470-be3f-28352f28bb48.png">

while

```
$ quarto render test-muy.qmd --to pdf
```

generates this PDF output:

<img width="874" alt="image" src="https://user-images.githubusercontent.com/285675/186194799-2e1c1699-d6ea-4a63-9d6e-0f46ededf245.png">

while

**Reply by mcanouil** [reply]
> Note that (as I learned yesterday from another quarto user), \newcommand does work.

I learned it today 😅
GitHub Issues and Discussion are great teaching resources 🤯

**Reply by cderv** [reply]
Wow cool ! Thanks @cscheid  ! 
This is a Pandoc feature right ? 

Or maybe just Mathjax in fact. Definition in one math block will be available to another block ? 🤔 That would explain. 
Anyway,  so cool! 🤯

**Reply by cscheid** [reply]
It's Pandoc - if you look at the generated HTML, it doesn't even have `\newcommand`. Pandoc is actually doing the macro replacement. Pretty incredible.

**Reply by cscheid** [reply]
(and thank you @eeholmes for that lesson!)

**Reply by cderv** [reply]
Wow indeed. Pretty incredible for sure ! And it seems it is there in Pandoc since a long time. Always happy to find unknown gems like that! 😄

**Reply by martintburgess** [reply]
Amazing, thank you to everyone who responded so quickly - I now how 4 different ways this will work (seems like my way was the only way that didn't 😅). But more importantly - I've learnt a tonne.
